File: retort-annotations/src/main/lombok/co/kaioru/retort/annotation/AnnotationFactory.java
package co.kaioru.retort.annotation;

import co.kaioru.retort.ICommand;
import co.kaioru.retort.ICommandContext;
import lombok.Getter;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;

@Getter
public class AnnotationFactory<I extends ICommandContext, O> implements IAnnotationFactory<I, O> {

    private final List<IAnnotationAdapter<I, O>> pipeline;

    public AnnotationFactory() {
        this.pipeline = new ArrayList<>();
    }

    @Override
    public List<ICommand<I, O>> generate(Object object) {
        List<ICommand<I, O>> commands = new ArrayList<>();

        for (Method method : object.getClass().getMethods()) {
            List<ICommand<I, O>> curCommands = new ArrayList<>();
            for (IAnnotationAdapter<I, O> a : getPipeline()) {
                if (a instanceof IAnnotationGenerator)
                    curCommands.add(((IAnnotationGenerator<I, O>) a).generate(object, method));
                if (a instanceof IAnnotationProcessor) {
                    curCommands.forEach(c -> {
                        ((IAnnotationProcessor<I, O>) a).process(c, object, method);
                        commands.add(c);
                    });
                }
            }
        }
        return commands;
    }

}

File: retort-annotations/src/main/kotlin/co/kaioru/retort/annotation/impl/AnnotationFactory.kt
package co.kaioru.retort.annotation.impl

import co.kaioru.retort.ICommand
import co.kaioru.retort.annotation.IAnnotationAdapter
import co.kaioru.retort.annotation.IAnnotationFactory
import co.kaioru.retort.annotation.IAnnotationGenerator
import co.kaioru.retort.annotation.IAnnotationProcessor
import co.kaioru.retort.annotation.exceptions.AnnotationAdapterException

open class AnnotationFactory<I, O> : IAnnotationFactory<I, O> {
    override val adapters: MutableCollection<IAnnotationAdapter<I, O>> = HashSet()

    override fun generate(input: Any): Collection<ICommand<I, O>> {
        val commands: MutableSet<ICommand<I, O>> = HashSet()

        input.javaClass.methods.forEach { method ->
            val currentCommands: MutableSet<ICommand<I, O>> = HashSet()

            adapters.forEach { adapter ->
                try {
                    if (adapter is IAnnotationGenerator)
                        currentCommands.add(adapter.generate(input, method))
                    if (adapter is IAnnotationProcessor)
                        currentCommands.forEach { adapter.process(it, input, method) }
                } catch (e: AnnotationAdapterException) {

                }
            }
            commands.addAll(currentCommands)
        }
        return commands
    }
}